beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -3209,34 +3319,75 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
                 curr_attestation_filter,
                 &self.spec,
             )
-            .map_err(BlockProductionError::OpPoolError)?
-            .into();
+            .map_err(BlockProductionError::OpPoolError)?;
         drop(attestation_packing_timer);
 
         let slot = state.slot();
         let proposer_index = state.get_beacon_proposer_index(state.slot(), &self.spec)? as u64;
 
-        let pubkey_opt = state
-            .validators()
-            .get(proposer_index as usize)
-            .map(|v| v.pubkey);
-
-        // Closure to fetch a sync aggregate in cases where it is required.
-        let get_sync_aggregate = || -> Result<SyncAggregate<_>, BlockProductionError> {
-            Ok(self
-                .op_pool
-                .get_sync_aggregate(&state)
-                .map_err(BlockProductionError::OpPoolError)?
-                .unwrap_or_else(|| {
-                    warn!(
-                        self.log,
-                        "Producing block with no sync contributions";
-                        "slot" => state.slot(),
-                    );
-                    SyncAggregate::new()
-                }))
+        let sync_aggregate = match &state {
+            BeaconState::Base(_) => None,
+            BeaconState::Altair(_) | BeaconState::Merge(_) => {
+                let sync_aggregate = self
+                    .op_pool
+                    .get_sync_aggregate(&state)
+                    .map_err(BlockProductionError::OpPoolError)?
+                    .unwrap_or_else(|| {
+                        warn!(
+                            self.log,
+                            "Producing block with no sync contributions";
+                            "slot" => state.slot(),
+                        );
+                        SyncAggregate::new()
+                    });
+                Some(sync_aggregate)
+            }
         };
```

Comment:
the total match is nice but we could also quite safely write this without enumerating all the forks

```suggestion
        let sync_aggregate = if matches!(&state, BeaconState::Base(_)) {
            None
        } else {
            self
                .op_pool
                .get_sync_aggregate(&state)
                .map_err(BlockProductionError::OpPoolError)?
                .unwrap_or_else(|| {
                    warn!(
                        self.log,
                        "Producing block with no sync contributions";
                        "slot" => state.slot(),
                    );
                    SyncAggregate::new()
                })
        };

```

Reply:
I often opt for the opposite - my reasoning being that when new forks are added, the code won't compile until your attention has been explicitly drawn to the places where forks have changed the logic. That way you can check to see if the fork has changed this again and if not you just make it:

`BeaconState::Altair(_) | BeaconState::Merge(_) | BeaconState::NewFork(_) ->`

But at least you made that decision *consciously*

What do you guys think?